lib/std/fmt.zig

```diff
@@ -1125,6 +1125,90 @@ pub fn formatIntBuf(out_buf: []u8, value: anytype, base: u8, uppercase: bool, op
     return fbs.pos;
 }
 
+/// Formats a duration in nanoseconds according to its magnitude:
+///
+/// - #ns
+/// - #[.#]us
+/// - #[.#]ms
+/// - #[.#]s
+/// - #m#[.#]s
+/// - #h#m#[.#]s
+/// - #d#h#m#[.#]s
+/// - #w#d#h#m#[.#]s
+pub fn formatDuration(ns: u64, writer: anytype) !void {
+    var ns_remaining = ns;
+
+    inline for (.{
+        .{ .ns = std.time.ns_per_week, .sep = 'w' },
+        .{ .ns = std.time.ns_per_day, .sep = 'd' },
+        .{ .ns = std.time.ns_per_hour, .sep = 'h' },
+        .{ .ns = std.time.ns_per_min, .sep = 'm' },
+    }) |unit| {
+        // Compare to total to retain intermediate units
+        if (ns >= unit.ns) {
+            const units = ns_remaining / unit.ns;
+            ns_remaining -= units * unit.ns;
+            try format(writer, "{}{c}", .{ units, unit.sep });
+        }
+    }
+
+    // Mutually exclusive
+    inline for (.{
+        .{ .ns = std.time.ns_per_s, .sep = "s" },
+        .{ .ns = std.time.ns_per_ms, .sep = "ms" },
+        .{ .ns = std.time.ns_per_us, .sep = "us" },
+    }) |unit| {
+        if (ns >= unit.ns) {
+            const units = @intToFloat(f64, ns_remaining) / @intToFloat(f64, unit.ns);
+            return format(writer, "{d}{}", .{ units, unit.sep });
```

Comment:
`d` will print the fp value with full precision, is that wanted?
Also, `{s}` for string formats please.

Reply:
It's probably reasonable to limit the max precision to 3 decimal places (i.e. down to the next unit). I don't think there's any way to do max precision with `fmt`, but I can `div(floor(mul))`

Reply:
You can set the [precision](https://github.com/ziglang/zig/blob/a1ba3c5aa435f4f36f3557e4732bf99d447eaf05/lib/std/fmt.zig#L46) parameter for that.

Reply:
Unfortunately `@as(f64, 1)` formatted with `{d:.3}` comes out as `1.000`, where I would prefer to omit unnecessary precision

Reply:
Is using floats here really necessary? Since `ns_remaining` is a fixed-point number, you can manually split it into integral and fractional parts, and print them with `formatInt`.

Reply:
Is eliminating the floating point operation worth the extra complexity?

```zig
-                // Floor to 3 decimal places
                 const kunits = ns * 1000 / unit.ns;
-                const units = @intToFloat(f64, kunits) / 1000.0;
-                try formatFloatDecimal(units, .{}, writer);
+                try formatInt(kunits / 1000, 10, false, .{}, writer);
+                // Write up to 3 decimal places
+                const frac = kunits % 1000;
+                if (frac > 0) {
+                    try writer.writeByte('.');
+                    if (frac % 10 > 0) {
+                        try formatInt(frac, 10, false, .{ .fill = '0', .width = 3 }, writer);
+                    } else if (frac % 100 > 0) {
+                        try formatInt(frac / 10, 10, false, .{ .fill = '0', .width = 2 }, writer);
+                    } else {
+                        try formatInt(frac / 100, 10, false, .{ .fill = '0', .width = 1 }, writer);
+                    }
+                }
```

Serious question, I don't often work in environments where 2 floating operations would be noticable

Reply:
Not sure how performance compares, but another option:

```zig
                const frac = kunits % 1000;
                if (frac > 0) {
                    try writer.writeByte('.');
                    var frac_buf: [3]u8 = undefined;
                    _ = formatIntBuf(&frac_buf, frac, 10, false, .{ .fill = '0', .width = 3 });
                    try formatBuf(frac_buf[0 .. std.mem.indexOfScalar(u8, &frac_buf, '0') orelse frac_buf.len], .{}, writer);
                }
```

Reply:
Thank you for the feedback! I have eliminated the floating point operations.